### DISCUSSION TITLE
Print a version where some of the functions are collapsed

### DISCUSSION THREAD

**Discussion by iilyak** [PDConSec/vsc-print#247] [Ideas]
When working on the codebase sometimes I need a 10000 feet view on the codebase. I can achieve that by collapsing all the functions defined in the class or module. Then I can expand one or two which I am interested in. I would like to be able to print out the result. 

Is it possible to implement this functionality?

**Comment by PeterWone** [comment]
@iilyak there are a number of things you'd need to do to accomplish this.

* You need to find out how to get code-folding state out of the editor
* You need to be able to print ranges of code with their line numbers being determined by their position in the underlying file

The second part I have already accomplished in support of printing selections. 

Getting the code-folding state out of the editor is not a documented operation and Microsoft was totally unhelpful when I asked for this information back in issue #53 where I proposed exactly this feature but eventually gave up on ever finding out how to get folding state. 

This information is certainly present in the source code which is publicly available. So whether it's possible depends on how determined you are. If you take this on as a research project and come back with details of how to get folding state, then I can and will use it. Or it would make a hell of a first issue. If you want to implement it yourself and raise a PR I'd be delighted to guide you through the methodology of printing ranges and composing a print job from fragments.

**Comment by iilyak** [comment]
I did a quick search in documentation and found following links

* [registerFoldingRangeProvider(selector: [DocumentSelector](https://code.visualstudio.com/api/references/vscode-api#DocumentSelector), provider: [FoldingRangeProvider](https://code.visualstudio.com/api/references/vscode-api#FoldingRangeProvider)): [Disposable](https://code.visualstudio.com/api/references/vscode-api#Disposable)]
* provideFoldingRanges(document: [TextDocument](https://code.visualstudio.com/api/references/vscode-api#TextDocument), context: [FoldingContext](https://code.visualstudio.com/api/references/vscode-api#FoldingContext), token: [CancellationToken](https://code.visualstudio.com/api/references/vscode-api#CancellationToken)): [ProviderResult](https://code.visualstudio.com/api/references/vscode-api#ProviderResult)<[FoldingRange](https://code.visualstudio.com/api/references/vscode-api#FoldingRange)[]>
* [folding.ts](https://github.com/microsoft/vscode/blob/6d45c3610aef49762dcf694bb598d47a0736ce38/extensions/typescript-language-features/src/languageFeatures/folding.ts#L19)
* [interface definition](https://github.com/microsoft/vscode/blob/6d45c3610aef49762dcf694bb598d47a0736ce38/src/vs/editor/common/languages.ts#L1411)

**Reply by PeterWone** [reply]
That's come a _long_ way since I last looked into it. Well done, on the face of it that's what I need. 

Be aware that I'm currently busy both with my day job and with re-engineering the Markdown rendering pipeline because the VS Code team blatantly disregards extension developer needs on anything that wasn't their own idea and things that used to work well are now broken for my purposes.

---

Now that I've had a chance to have a good look at those APIs, they don't support interrogating the state of folding, they provide a way to install your own folding provider (logic for identifying collapsible ranges) as demonstrated by this extension https://github.com/mtbaqer/vscode-better-folding.
